### ISSUE TITLE
dataclass wrapper causes "no arguments to constructor" type error

### ISSUE THREAD

**Issue by milisims** [microsoft/pyright#6292]
**Describe the bug**
I want to use dataclass with some default arguments as a decorator, but result removes the constructor arguments

**Code or Screenshots**
```python
from dataclasses import dataclass

component = dataclass() # want to wrap around dataclass, but this is minimal

@component
class Foo:
    bar: float = 0.

@dataclass()
class Bar:
    baz: float = 0.

f = Foo(bar=1)  # Expected no arguments to "Foo" constructor
b = Bar(baz=1)  # no type error
```

**VS Code extension or command-line**
Pyright as an lsp in neovim

**Comment by erictraut** [comment]
Static type checkers require that you apply a `dataclass` decorator directly to a class. You can't instantiate `dataclass` separately and later apply it to a class as you're doing here. That's because there are arguments to `dataclass()` that affect its behavior, and these arguments must be evaluated statically. They include parameters `frozen`, `init`, etc.

You may be able to cobble something together with `dataclass_transform` to do what you'd like. For details, refer to [PEP 681](https://peps.python.org/pep-0681/).

In any case, this isn't a bug. This is expected behavior, so I'm going to close the issue.

**Comment by milisims** [comment]
I'm not actually modifying any of the dataclass parameters, just using it for registration of components, but that still makes sense - thanks for the explanation
